### PR TITLE
Add Javadoc rules to Checkstyle without full enforcement #6582

### LINKED ISSUE #6582: Add Javadoc rules to Checkstyle without full enforcement
Quoting myself:

> Most attempts to enforce any Javadoc-related rule will result in "Curse it, 1000 violations? I can't do this alone."

Regardless, if we do not put any kind of enforcement, the Javadoc-related technical debt will only accumulate. Steps have to be taken to introduce these rules gradually.

### PR DESCRIPTION
**Pull request by wkurniawan07**
Fixes #6582 

**Outline of Solution**

Rules added:
- `JavadocMethod`, `JavadocType`, `JavadocVariable`, `JavadocPackage`: Javadoc has to be present in methods, classes, variables, and packages respectively.
- `AtclauseOrder`: `@param`, `@return`, etc. has to follow a specific order.
- `JavadocParagraph`: `<p>` is to be used for paragraphs after the first one that are separated by empty lines. Strictly to be placed before the first word of each new paragraph, with no whitespace or line break in between.
- `JavadocStyle`: Checks that the first sentence of the Javadoc ends with a period `.`.
- `NonEmptyAtclauseDescription`: `@param parameter` is not allowed; `@param parameter explanation of the parameter` is. Combined with `JavadocMethod`'s capability of enforcing `@param`, we get to enforce existence AND explanation for all method parameters' documentation.
- `JavadocTagContinuationIndentation`: 8 spaces indentation for line breaks in at clauses.
- `SingleLineJavadoc`: Prohibits at clauses in single line Javadoc.

**Enforcement Level**

- `JavadocMethod`, `JavadocType`, `JavadocVariable`: enforced only for `public` methods in very limited packages only. We already have an overwhelming debt for public methods so let's not talk about anything less visible (`protected` etc.).
- `NonEmptyAtclauseDescription`: not enforced at all.

All other rules are fully enforced as long as the Javadoc is present.

**Fixes included**

- Added `package-info.java` in all packages with content lifted from `design.md`.
- Some methods and variables are downgraded from `public` to `private` to "avoid" the rules enforcement. This is like sweeping the problem under the rug, but the fact that those methods are unnecessarily `public` already speaks of another problem, and when the method is `private`, a Javadoc is much less important.
- Fixed phrases like this in Javadoc: `HashMap containing <class name, package name>`. `< ... >` is parsed as HTML and thus invalid contents inside will cause parsing failure.
- Fixed some invalid HTML such as unclosed `<a>` and `<code>` tag.
- Fixed some misused Javadoc: explaining "how" instead of "what", mentioning "used by".
- Some unused methods are removed to reduce the maintenance effort.

**What I did not do**

Checking the validity of the Javadocs being touched in this PR.

### CODE REVIEW

**Comment by xpdavid** [comment]
@wkurniawan07 It this PR completed?

**Comment by wkurniawan07** [comment]
@xpdavid still experimenting, wait for the `s.ToReview` label.

**Comment by wkurniawan07** [comment]
@xpdavid can review now. If you feel that some rules are unnecessarily restrictive, feel free to discuss.

I also tried running "Generate Javadoc..." from Eclipse for all visibility levels and found (and fixed) some warnings it generated. As it turns out, `@see` and `@throws` automatically links the entry and thus there is no need to `@link` for those tags, something that I just learned as well. On top of that, the warning comes from wrong parameter name documentation (likely because of copy-paste error, or the method is modified but the doc is not), which is actually enforce-able via CheckStyle but haven't been done yet.

**Comment by xpdavid** [comment]
@wkurniawan07 Will review soon. Sorry for the late review.

File: src/test/java/teammates/test/pageobjects/AppPage.java [hunk 1]
```diff
@@ -803,8 +788,9 @@ public void verifyTablePattern(int tableNum, int column, String patternString) {
      * Verifies that the currently loaded page has the same HTML content as
      * the content given in the file at {@code filePath}. <br>
      * The HTML is checked for logical equivalence, not text equivalence.
-     * @param filePath If this starts with "/" (e.g., "/expected.html"), the
-     * folder is assumed to be {@link TestProperties.TEST_PAGES_FOLDER}.
+     * @param filePath
+     *         If this starts with "/" (e.g., "/expected.html"), the
```
Comment (xpdavid) [inline src/test/java/teammates/test/pageobjects/AppPage.java]: About the separation in `@param` between parameter name and description. To be consistent, are we using a line separator or hyphen or just a space?

File: src/client/java/teammates/client/scripts/DataGenerator.java [hunk 2]
```diff
@@ -11,56 +11,47 @@
...
- * of the objects to save memory.
- * 
- * Second - To make this script portable, if you remove the teammates.common.Common import, and define
- * the path yourself, this become a totally portable script.(this speeds up the process of creating data)
- * 
- * Third - The format of data bundle is json but quite "strange", no relationships, no arrays.
- * 
+ * <p>This script does not use any teamamtes's data structures or json framework for some reasons:
+ * <ul>
+ * <li>For 5000 or more students, it will consume a lot of memory. Need to store only id
```
Comment (wkurniawan07) [inline src/client/java/teammates/client/scripts/DataGenerator.java]: We have enough rules to make up a huge amount of technical debt. Let's not create another. This is not a big issue anyway.

File: src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java [hunk 3]
```diff
@@ -915,7 +915,7 @@ private String getNameFromRoster(String participantIdentifier, boolean isFullNam
     /**
      * Get the displayable full name from an email.
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java]: I don't even touch this line... 😑

File: src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java [hunk 4]
```diff
@@ -915,7 +915,7 @@ private String getNameFromRoster(String participantIdentifier, boolean isFullNam
...
      * 
-     * This function is different from getNameForEmail as it obtains the name
+     * <p>This function is different from {@link #getNameForEmail} as it obtains the name
      * using the class roster, instead of from the responses.
      * @param participantIdentifier
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java]: The `JavadocParagraph` rule claims that it can enforce such, but obviously it's buggy. Perhaps future versions of Checkstyle (7.x) has fixed it.
(if you're wondering why we're stuck at version 6.19, it's because of Java 8)

File: src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java [hunk 5]
```diff
@@ -1806,7 +1804,7 @@ public String getQuestionText(String feedbackQuestionId) {
     }
     
     /**
-     * Returns responses as a Map<recipientName, Map<question, List<response>>>
+     * Returns responses as a {@code Map<recipientName, Map<question, List<response>>>}
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java]: > Should end with a period.

Nope. Take a look at the next line.

File: src/main/java/teammates/common/datatransfer/InstructorPrivileges.java [hunk 6]
```diff
@@ -283,8 +283,6 @@ private void updatePrivilegeInSessionLevel(String sectionName, String sessionNam
     }
     
     /**
-     * used for bulk update of privileges for sectionName
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/InstructorPrivileges.java]: This question that has no single answer. If you ask me, as long as the `@` clauses do enough explanation, e.g. the following:
```java
/**
 * @param n1 the first number
 * @param n2 the second number
 * @return the sum of the two numbers
 */
public void getSum(int n1, int n2) {
    return n1 + n2;
}
```
Then adding a description does not value add and becomes a repetition.

Of course, you can argue the other way around:
```java
/**
 * Gets the sum of two numbers passed as parameters.
 * 
 * @param n1 the first number
 * @param n2 the second number
 */
public void getSum(int n1, int n2) {
    return n1 + n2;
}
```
But the thing is Checkstyle *does* have a check for missing `@return` tag.

File: src/main/java/teammates/common/datatransfer/attributes/FeedbackSessionAttributes.java [hunk 7]
```diff
@@ -346,7 +346,7 @@ public boolean isClosingWithinTimeLimit(int hours) {
     }
     
     /**
-     * returns true if the session is closed within the past hour of calling this function
+     * @return true if the session is closed within the past hour of calling this function
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/attributes/FeedbackSessionAttributes.java]: Sadly you have been proven wrong, since `SingleLineJavadoc` is already in action.

File: src/main/java/teammates/storage/api/ProfilesDb.java [hunk 8]
```diff
@@ -49,12 +49,11 @@ public StudentProfileAttributes getStudentProfile(String accountGoogleId) {
      * Assumes that the googleId remains the same and so updates the profile
      * with the given googleId.
      * 
-     * TODO: update the profile with whatever given values are valid and
-     * ignore those that are not valid.
      * @param newSpa
      * @throws InvalidParametersException
      * @throws EntityDoesNotExistException
      */
+    // TODO: update the profile with whatever given values are valid and ignore those that are not valid.
```
Comment (wkurniawan07) [inline src/main/java/teammates/storage/api/ProfilesDb.java]: We've already enforced `EmptyLineSeparator`, haven't we?

File: src/main/java/teammates/ui/controller/AdminInstructorAccountAddAction.java [hunk 9]
```diff
@@ -286,19 +282,21 @@ private String getDemoCourseIdRoot(String instructorEmail) {
...
-    */
+     * Generate a course ID for demo course from a given email or a generated course Id.
+     * 
+     * <p>Here we check the input string is an email or course Id and handle them accordingly;
+     * check the resulting course id, and if bigger than maximumIdLength, cut it so that it equals maximumIdLength.
+     * 
+     * @param instructorEmailOrProposedCourseId is the instructor email or a proposed course id that already exists.
+     * @param maximumIdLength is the maximum resulting id length allowed, above which we will cut the part before "@"
+     * @return the proposed course id, e.g.:
+     *         <ul>
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/controller/AdminInstructorAccountAddAction.java]: Same answer.

File: src/main/java/teammates/ui/controller/ImageResult.java [hunk 10]
```diff
@@ -15,12 +15,14 @@
...
 public class ImageResult extends ActionResult {
 
+    /** The Google Cloud Storage blob key for the image. */
     public String blobKey;
     
     public ImageResult(String destination, String blobKey, AccountAttributes account,
             List<StatusMessage> status) {
         super(destination, account, status);
         this.blobKey = blobKey;
+        // a
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/controller/ImageResult.java]: Oops. Not sure where this comes from, but will fix.

File: src/main/java/teammates/common/datatransfer/attributes/FeedbackSessionAttributes.java (hunk 7 above)
Comment (xpdavid) [inline src/main/java/teammates/common/datatransfer/attributes/FeedbackSessionAttributes.java]: I thought it is 
`/** @return bla bla bla */`

Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/attributes/FeedbackSessionAttributes.java]: What you showed is precisely what `SingleLineJavadoc` will complain about...

File: src/test/java/teammates/test/pageobjects/AppPage.java (hunk 1 above)
Comment (wkurniawan07) [inline src/test/java/teammates/test/pageobjects/AppPage.java]: Re-quoting myself:
We have enough rules to make up a huge amount of technical debt. Let's not create another. This is not a big issue anyway.

**Comment by wkurniawan07** [comment]
> Let's improve the documentation step by step.

Indeed, this is the only way. It may take 3 months, 6 months, one year, but we will eventually get there.

<sub>Or if you have a team doing nothing but adding documentation, it might be finished in a month. Just like how all the CheckStyle etc. rules are put in place. 🤓</sub>

File: src/main/java/teammates/storage/api/StudentsDb.java [hunk 11]
```diff
@@ -263,8 +259,6 @@ public StudentAttributes getStudentForRegistrationKey(String registrationKey) {
     }
 
     /**
-     *  Preconditions: <br>
-     *  All parameters are non-null
```
Comment (damithc) [inline src/main/java/teammates/storage/api/StudentsDb.java]: Why is this removed? It's part of the contract enforced by the method.

File: src/main/java/teammates/ui/controller/AdminInstructorAccountAddAction.java [hunk 12]
```diff
@@ -268,11 +264,11 @@ private String generateDemoCourseId(String instructorEmail) {
     }
     
     /** 
-    * Generate a course ID for demo course from a given email
-    * 
-    * @param instructorEmail is the instructor email.
-    * @return the first proposed course id. eg.lebron@gmail.com -> lebron.gma-demo
-    */
+     * Generate a course ID for demo course from a given email.
```
Comment (damithc) [inline src/main/java/teammates/ui/controller/AdminInstructorAccountAddAction.java]: `Generates`?

File: src/client/java/teammates/client/scripts/PerformanceProfiler.java [hunk 13]
```diff
@@ -118,10 +120,10 @@ public void run() {
     }
     
     /**
-     * This function perform the method and print the return value for debugging
+     * This function perform the method and print the return value for debugging.
      * @param method
```
Comment (wkurniawan07) [inline src/client/java/teammates/client/scripts/PerformanceProfiler.java]: Not enforced yet. Enforcing it will easily invoke 5000+ violations.

File: src/client/java/teammates/client/scripts/PerformanceProfiler.java [hunk 14]
```diff
@@ -158,19 +160,16 @@ public void performMethod(Method method) {
...
         new PerformanceProfiler(defaultReportPath).start();
     }
 
     /**
-     * The results from file stored in filePath
+     * The results from file stored in filePath.
      * @param filePath
-     * @return HashMap<nameOfTest,durations> of the report stored in filePath
+     * @return {@code HashMap<nameOfTest, durations>} of the report stored in filePath
      * @throws IOException
```
Comment (wkurniawan07) [inline src/client/java/teammates/client/scripts/PerformanceProfiler.java]: Yes, the same way empty `@param` and `@return` are enforced.

File: src/main/java/teammates/common/datatransfer/InstructorPrivileges.java (hunk 6 above)
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/InstructorPrivileges.java]: As it so happens, I found a way for CS to disallow empty description.

But this results in a dilemma:
- Disallowing both empty description and empty `@return` tag necessitates duplication.
- Disallowing empty `@return` tag makes documenting legitimate, non-trivial return values missable.
- Disallowing empty description is against the coding standard for return statements.

For me disallowing empty `@return` is still the best alternative since "coding standard" is not a bible.

File: src/main/java/teammates/common/datatransfer/attributes/FeedbackSessionAttributes.java (hunk 7 above)
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/attributes/FeedbackSessionAttributes.java]: This comes back to the previous discussion of `@return` vs `Returns`.

**Comment by wkurniawan07** [comment]
> Going forward, I'm not sure if we want to enforce `@param` and `@return` for all public methods as there can be cases where there's nothing more to say about a parameter or a return value.

The problem here is that CS has no concept of "partial enforcement". Either we shift the burden of checking to the reviewers, or we allow seemingly redundant documentation. Admittedly, none of the alternative sound good.

> However, we should block empty `@param` and `@return` as they don't add any value.

In CS' rule, this is independent of presence of `@param` and `@return` and so is doable, although not enforced yet.

File: src/main/java/teammates/common/datatransfer/InstructorPrivileges.java (hunk 6 above)
Comment (damithc) [inline src/main/java/teammates/common/datatransfer/InstructorPrivileges.java]: > For me disallowing empty @return is still the best alternative since "coding standard" is not a bible.

Did you mean you prefer this? That's my preference too, but I'm not sure if that's what you meant.

```java
/**
 * Gets the sum of two numbers passed as parameters.
 * 
 * @param n1 the first number
 * @param n2 the second number
 */
public void getSum(int n1, int n2) {
    return n1 + n2;
}
```

Regarding `Returns` vs `@returns`, my initial intuition was to prefer `@return` as it is more formal. However, the Google coding standard gives a legit reason for preferring `Returns` which made me change my preference as well. The reason given is that in some contexts, overview statement is the only documentation shown and therefore it takes priority over `@return`.

**Comment by xpdavid** [comment]
@wkurniawan07 BTW, if we want to implement the `partial enforcement`, it's doable.
Here is some part in CheckStylee `JavadocMethodCheck`
``` java
// ...
                checkParamTags(tags, ast, reportExpectedTags);
                checkThrowsTags(tags, getThrows(ast), reportExpectedTags);
                if (CheckUtils.isNonVoidMethod(ast)) {
                    checkReturnTag(tags, ast.getLineNo(), reportExpectedTags);
                }
// ...
```
We can write conditional statement for `params` check, `throws` check and `return` check.

**Comment by wkurniawan07** [comment]
@xpdavid well if you want to open up the black box, anything is possible... 😛

**Comment by wkurniawan07** [comment]
@xpdavid by the way, by "partial enforcement" I mean enforcing the same rule for some code block but not others. For example:
```java
public void getSum(int n1, int n2) {
    return n1 + n2;
}

public String someComplicatedMethod(int whatParamIsThis, String andThis) {
    // do something complicated
}
```
We can reason to skip the docs for the first method and not the second, but CS will not. We do have the `// CHECKSTYLE.OFF` workaround but over-using it will make the code cluttered.

**Comment by wkurniawan07** [comment]
@xpdavid @damithc 

So I tweaked some rules, particularly:
- `JavadocMethod`: allow missing `@param` and `@return` tag but not `@throws` tag. Still, not enforced yet.
- `JavadocStyle`: disallow empty description. Enforced for all.
- `NonEmptyAtclauseDescription`: enforced for all. Things like `@param parameter` and `@return` without further description are removed because they are not meaningful anyway.

File: src/main/java/teammates/logic/core/FeedbackSessionsLogic.java [hunk 15]
```diff
@@ -201,8 +196,9 @@ public FeedbackSessionAttributes copyFeedbackSession(String newFeedbackSessionNa
     }
 
     /**
+     * Returns true if there is some open or published email sent for the course.
+     *
```
Comment (xpdavid) [inline src/main/java/teammates/logic/core/FeedbackSessionsLogic.java]: Unnecessary line?

File: src/main/java/teammates/storage/api/StudentsDb.java [hunk 16]
```diff
@@ -263,9 +260,7 @@ public StudentAttributes getStudentForRegistrationKey(String registrationKey) {
...
```
Comment (xpdavid) [inline src/main/java/teammates/storage/api/StudentsDb.java]: From previous review by @damithc 
> Why is this removed? It's part of the contract enforced by the method.

File: src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java [hunk 17]
```diff
@@ -28,7 +28,7 @@
  * <br> * The basic {@link FeedbackSessionAttributes}
  * <br> * {@link List} of viewable responses as {@link FeedbackResponseAttributes} objects.
  */
-public class FeedbackSessionResultsBundle implements SessionResultsBundle {
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java]: Unused interface (just take a look, it's an empty interface), and I'm not going to maintain the docs for something unused.

File: src/main/java/teammates/common/util/ActivityLogEntry.java [hunk 18]
```diff
@@ -285,15 +286,15 @@ public String getIconRoleForShow() {
 
     /**
      * Assumption: the {@code requestUrl} is in the format "/something/actionName"
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/util/ActivityLogEntry.java]: There is no such thing in Javadoc.

File: src/main/java/teammates/logic/core/AccountsLogic.java [hunk 19]
```diff
@@ -225,10 +218,6 @@ private void confirmNotRejoiningUsingDifferentKey(
 
     }
 
-    /**
-     * @throws JoinCourseException if the instructor has already joined this
```
Comment (wkurniawan07) [inline src/main/java/teammates/logic/core/AccountsLogic.java]: Two reasons:
1. The method name has already suggested what it is going to do.
1. If this remains, `JavadocStyle` will complain because the description is empty, and I can't really craft a document that doesn't make the `@throws` description redundant.

File: src/main/java/teammates/logic/core/FeedbackSessionsLogic.java (hunk 15 above)
Comment (wkurniawan07) [inline src/main/java/teammates/logic/core/FeedbackSessionsLogic.java]: Don't we agree that there should be an empty line between description and at clauses (although not enforced)?
